CUMULUS-2255: Upgrade supported TF version to 0.14.4

markdboyd opened the pull request:
**Summary:** Summary of changes

Addresses [CUMULUS-2255: Upgrade supported TF version to 0.14.4](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2255)

Corresponding PR for the template-deploy repo: https://github.com/nasa/cumulus-template-deploy/pull/57

## Changes

* Update all TF code for compatibility with TF version 0.13.6

**Testing notes:** see the `upgrading-tf-version-0.13.6.md` document

## PR Checklist

- [x] Update CHANGELOG
- [x] Write up any user facing docs for the upgrade
- [x] Manual testing

npauzenga commented on docs/deployment/README.md:
```diff
@@ -52,13 +52,13 @@ For other cases,
- $ tfenv install 0.12.12
-[INFO] Installing Terraform v0.12.12
+ $ tfenv install 0.14.4
+[INFO] Installing Terraform v0.14.4
```
From the ticket it looks like the intention here is to use features available in Terraform v0.13 and greater (and that's what we're now requiring in our modules below). I'm wondering if it would make more sense to instruct users to use 0.13.6 (last patch version of 0.13) here instead of the latest version, 0.14.4?

My concern is mainly around having users on the same page. There shouldn't be breaking changes between minor versions so technically 0.13 and 0.14 should be interchangeable but I'm not sure if there would be noticeable differences that could confuse users. 

Or, if we're mostly testing this with 0.14.4, maybe the required version should be >=0.14?

markdboyd replied:
From what I read, the ticket says that they want to be able to use more recent versions of Terraform. At the time of writing, 0.14 was only in beta, but now it has stable releases so it seems like we should upgrade to support it, if possible.

And indeed, in my testing, there are only breaking changes between 0.12 and 0.13, not 0.13 and 0.14. Most of the changes in this PR involve updating our Terraform code for compatibility with > 0.13. 

I'm not sure if that answers the question?

markdboyd replied:
I take it all back. We had to make the upgrade only to 0.13.6 for now because for various reasons, you can't easily upgrade from 0.12.12 straight to 0.14.4

Jkovarik commented on docs/deployment/README.md:
````diff
@@ -52,13 +52,13 @@ For other cases,
 ```shell
- $ tfenv install 0.12.12
-[INFO] Installing Terraform v0.12.12
+ $ tfenv install 0.14.4
````
Worth noting: 14.5 is out.

markdboyd replied:
Resolving since we're not going to do the upgrade to 0.14

Jkovarik commented:
Running init against an existing v12 stack with remote state (but no local)  direct to v13 results in : 

```
terraform init
Initializing modules...

Initializing the backend...

Error: Invalid legacy provider address

This configuration or its associated state refers to the unqualified provider
"aws".

You must complete the Terraform 0.13 upgrade process before upgrading to later
versions.
```

Jkovarik commented:
Re: rolling back -- Presumably rolling back also involves updating the dynamo table?

Jkovarik commented:
Running a terraform init --reconfigure results in:

```
terraform init --reconfigure
Initializing modules...
- data_persistence in ../../tf-modules/data-persistence

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.

Error: Invalid legacy provider address

This configuration or its associated state refers to the unqualified provider
"random".

You must complete the Terraform 0.13 upgrade process before upgrading to later
versions.


Error: Invalid legacy provider address

This configuration or its associated state refers to the unqualified provider
"aws".

You must complete the Terraform 0.13 upgrade process before upgrading to later
versions.
```

Jkovarik commented:
We should include updated notes re: 12->13 provider updates.  Also worth noting, the terraform 0.13upgrade isn't available under terraform 14 (it's been removed), so some pre-work needs to be done under terraform v12.

Jkovarik reviewed (changes requested):
The changes mostly look good -- I was able to confirm/deploy up to 13 and 14 for data persistence. 

From our discussion -- this needs: 

* To be only an update to use only 0.13.6 as it's not possible to move 12->14 due to remote state updates
* A user-doc describing the deploy changes, including references to update docs and what they might need to do for template-deploy 
* A warning (CHANGELOG? )that running `apply` will change their statefile without requiring approval

We also need to make sure this gets backported to any maintenance branches so as to avoid needing to swap terraform versions in CI, and should notify partners that's going to happen as it might require work on their part.  

We *also* need to put a PR in against feature-rds to make sure the terraform there is v13 compatible to avoid terraform version swapping for that branch. 

I'll test the Core deployment for 14 later this afternoon.

markdboyd commented:
@Jkovarik 

> Re: rolling back -- Presumably rolling back also involves updating the dynamo table?

I'm not sure what this means?

> A warning (CHANGELOG? )that running apply will change their statefile without requiring approval

I'm also not sure what this refers to? 

> We also need to make sure this gets backported to any maintenance branches so as to avoid needing to swap terraform versions in CI, and should notify partners that's going to happen as it might require work on their part.

Do you mean maintenance branches for 5.0.x? None currently exist, correct?

> We also need to put a PR in against feature-rds to make sure the terraform there is v13 compatible to avoid terraform version swapping for that branch.

Agreed, but we can do that once this is merged to master and we're going to take master back to feature-rds. I can do it as an immediate follow-up

Otherwise, I have:

- Updated the code to only perform the update to 0.13.6 and updated all doc references
- Adding user facing documentation for the upgrade

Jkovarik commented:
@markdboyd  

> Re: rolling back -- Presumably rolling back also involves updating the dynamo table?

That was referring to the dev/testing instructions.   It *does* - you need to manually update dynamo to roll back.    More for posterity/new devs than a comment on the PR itself. 

> A warning (CHANGELOG? )that running apply will change their statefile without requiring approval

This is suggesting we note in the instructions that the v 13 terraform apply will update the remote statefile from v12 to v13 without requiring approval if there are no changes. 

> We also need to make sure this gets backported to any maintenance branches so as to avoid needing to swap terraform versions in CI, and should notify partners that's going to happen as it might require work on their part.

I think we need to put the v15 release branch on v13, or come up with another solution for backports as CI stacks can't automatically go back and forth between TF revisions.

Jkovarik reviewed (approved):
Approving with a couple of comments/nits.   Nice work!

markdboyd commented on docs/upgrade-notes/upgrading-tf-version-0.13.6.md:
```diff
@@ -0,0 +1,42 @@
+
+### Prerequisites
+
+- Follow the [Terraform guidance for what to do before upgrading](https://www.terraform.io/upgrade-guides/0-13.html#before-you-upgrade), notably ensuring that you have no pending changes to your Cumulus deployments before proceeding.
```
https://github.com/nasa/cumulus/pull/2048/commits/9dc8dcef0c567231090741041f4968cfbfa636f8

markdboyd commented on docs/upgrade-notes/upgrading-tf-version-0.13.6.md:
````diff
@@ -0,0 +1,42 @@
+    ```
+
+3. Run `terraform apply` to perform a deployment.
+4. Repeat steps 2 and 3 for the `cumulus-tf` directory.
````
https://github.com/nasa/cumulus/pull/2048/commits/9dc8dcef0c567231090741041f4968cfbfa636f8

markdboyd commented on docs/upgrade-notes/upgrading-tf-version-0.13.6.md:
```diff
@@ -0,0 +1,42 @@
+
+### Prerequisites
+
+- Follow the [Terraform guidance for what to do before upgrading](https://www.terraform.io/upgrade-guides/0-13.html#before-you-upgrade), notably ensuring that you have no pending changes to your Cumulus deployments before proceeding.
```
https://github.com/nasa/cumulus/pull/2048/commits/aee622a356bcb6edcbaa14b7d158bfe74e176a02
